### PR TITLE
button spec

### PR DESCRIPTION
**Pull request by pelikhan**
PR to track changes. Feel free to comment, edit. @tballmsft 

Here are the related PRs: https://github.com/microsoft/jacdac-c/pull/15, https://github.com/microsoft/jacdac-ts/pull/181

### CODE REVIEW

**Review by tballmsft — COMMENTED** [review]
Here's my suggestion for hold_time. Also, we agreed to add time to up event. I will do that next.

File: services/button.md [hunk 1]
```diff
@@ -23,9 +23,10 @@ The amount of time a button needs to be held for a `hold` event to be generated.
 
 Emitted when button goes from inactive (`pressed == 0`) to active.
 
-    event up @ inactive
+    event up { time: u32 ms } @ inactive
```
Comment (pelikhan) [inline services/button.md:26]: i would say that u16 is enough here. where do we want to cap the time? how much can padauk handle?

We also have the 8ms unit if needed.

File: services/button.md [hunk 2]
```diff
@@ -13,24 +13,26 @@ A push-button, which returns to inactive position when not operated anymore.
...
 
 Emitted together with `up` when the press time was not longer than 500ms.
 
-    event long_click @ 0x81
-
-Emitted after button is held for 500ms. Long click events are followed by a separate up event.
+    event hold @ 0x81
 
-    event hold @ 0x82
+Emitted after the button is held for more than `hold_time' (default/minimum 500ms). Hold events are followed by a separate up event.
```
Comment (pelikhan) [inline services/button.md:37]: ``
(default/minimum 500ms)
``

that's defined in the ``hold_time`` spec so i woudl not repeat here.

File: services/button.md [hunk 3]
```diff
@@ -13,24 +13,28 @@ A push-button, which returns to inactive position when not operated anymore.
 
 Indicates whether the button is currently active (pressed).
 
+    rw hold_time: u16 { absolute_min=500 } ms @ 0x80
```
Comment (mmoskal) [inline services/button.md:16]:
```suggestion
    rw hold_time = 1000: u16 { absolute_min=500 } ms @ 0x80
```

Comment (mmoskal) [reply services/button.md:16]: also the 500 absolute min may be a bit low

**Comment by mmoskal** [comment]
I think we should use `us` not `ms` for all time units - I think Padauk will only keep time in us not in ms. @jamesadevine  ?

File: services/button.md (hunk 3 above)
Comment (pelikhan) [reply services/button.md:16]: i would say at least 1500 for hold, which allows for longer holds.

**Comment by mmoskal** [comment]
though I guess we could keep time on padauk in 4us units, and pretend a ms is 1024us?

**Comment by pelikhan** [comment]
``o`` -> 1024 || ``ous`` -> ``1024us`` ?

File: services/button.md (hunk 3 above)
Comment (mmoskal) [reply services/button.md]: sorry, actually I meant too restrictive; maybe 100ms

**Comment by tballmsft** [comment]
so then we need u32

**Comment by tballmsft** [comment]
OK. How does it look now?

**Comment by mmoskal** [comment]
I think we're still not sure if we want us or ms...

**Comment by tballmsft** [comment]
agreed. I would say we should differentiate sensors dependent on human response time from those that are not dependent on human. For the former, ms. For the latter, us.

**Comment by jamesadevine** [comment]
>  I think Padauk will only keep time in us not in ms

Padauk records time passed by a 6 ms timer interrupt!

**Comment by tballmsft** [comment]
I reverted to ms and changed to one threshold (per @mmoskal's suggestion).

**Comment by tballmsft** [comment]
I am finished implementing spec, code, and tests. All good and ready to submit.

**Comment by pelikhan** [comment]
Can you link all the PRs together? Pasting URLs in the description is enough

File: services/button.md [hunk 4]
```diff
@@ -13,24 +13,27 @@ A push-button, which returns to inactive position when not operated anymore.
...
+records the amount of time between the down and up events.
 
     event click @ 0x80
 
-Emitted together with `up` when the press time was not longer than 500ms.
-
-    event long_click @ 0x81
-
-Emitted after button is held for 500ms. Long click events are followed by a separate up event.
+Emitted together with `up` when the press time is not longer than `click_hold_time`.
```
Comment (pelikhan) [inline services/button.md:35]: not longer -> less than?

**Comment by pelikhan** [comment]
LGTM

**Comment by tballmsft** [comment]
Here are the related PRs: https://github.com/microsoft/jacdac-c/pull/15, https://github.com/microsoft/jacdac-ts/pull/181

**Pull request merged**